src/Language/Futhark/Unused.hs

```diff
@@ -0,0 +1,207 @@
+module Language.Futhark.Unused (findUnused, getBody, getDecs) where
+
+import Data.Bifunctor qualified as BI
+import Data.Foldable (Foldable (foldl'))
+import Data.List.NonEmpty qualified as NE
+import Data.Map.Strict qualified as M
+import Data.Maybe (catMaybes, maybeToList)
+import Data.Set qualified as S
+import Language.Futhark
+import Language.Futhark.Semantic (FileModule (FileModule), includeToFilePath)
+
+-- Steps:`
+-- ∘ For each definition of a VName x in any file (corresponding to a top level definition), find which VNames are mentioned in the definition of x. (Intuitively, which functions are called directly.)
+-- ∘ Compute the transitive closure of the information gathered in (1), which means you now have a mapping from each VName to the set of VNames used in its definition (even indirectly).
+-- ∘ For those VNames defined in root files, take the union of their reachable names.
+-- ∘ Subtract the set produced in (3) from the set of all VNames. This gives you a set of VNames that are not reachable from the root files.
+
+-- Guidelines:
+-- (1) create a mapping for functions in *any* import to the directly usable functions in it.
+-- (2) create a mapping from normalized filename to top-level functions inside the filename.
+-- (3) Compute transitive closure for the map in (1)
+
+-- We are looking for VNames that represent functions.
+
+-- cabal run futhark -- unused un/b.fut
+
+getDecs :: FileModule -> [DecBase Info VName]
+getDecs (FileModule _ _env (Prog _doc decs) _) = decs
+
+findUnused :: [FilePath] -> [(ImportName, FileModule)] -> M.Map FilePath [(VName, SrcLoc)]
```

Comment:
You should add a comment for what this `[FilePath]` means.  If I recall correctly, it's the root reachable files or something like that.